refnx/reduce/platypusnexus.py

```diff
@@ -626,6 +627,118 @@ def _check_sample_environments(self, d, h5d):
         return d
 
 
+class SpinChannel(Enum):
+    """
+    Describes the spin state of a polarised neutron beam.
+    """
+
+    UPUP = (1, 1)
+    UPDOWN = (1, 0)
+    DOWNUP = (0, 1)
+    DOWNDOWN = (0, 0)
+
+
+class SpinSet(object):
+    """
+    Describes a set of spin-channels at a given angle.
+
+    TODO: implement polarisation efficiency correction within this class
+
+    Parameters
+    ----------
+    list of {str, h5data}
+        list of str, or list of h5py file handles pointing to
+        a set of polarised neutron beam files
+
+    data_folder: {str, Path}
+        Path to the data folder containing the data to be reduced.
+    """
+
+    def __init__(self, dataset, data_folder=None):
+
+        # Currently only Platypus has polarisation elements
+        self.reflect_klass = PlatypusNexus
+
+        self.data_folder = data_folder
+        if len(dataset) > 4:
+            raise ValueError("Too many spin channels!")
+        if len(dataset) < 2:
+            raise ValueError(
+                "Too few spin channels! Need at least R++ and R--."
+            )
+        if isinstance(dataset[0], PlatypusNexus):
+            self.beams = self.arrange_spin_channels(dataset)
+        elif type(dataset[0]) is str:
+            fnames = [os.path.join(self.data_folder, dset) for dset in dataset]
+            PLPset = [self.reflect_klass(d) for d in fnames]
+
+            self.beams = self.arrange_spin_channels(*PLPset)
+        else:
+            raise ValueError("Dataset should be list of PlatypusNexus or str")
+        self.mm, self.mp, self.pm, self.pp = self.beams
+
+    @property
+    def spin_channels(self):
+        return [
+            s.spin_state.value if s is not None else None
+            for s in [self.mm, self.mp, self.pm, self.pp]
+        ]
+
+    def arrange_spin_channels(self, *data):
+        """
+        Function that takes a random input of spin channels at
+        a single angle and returns them arranged according to
+        --, -+, +-, ++ and replaces the spin channel with None
+        if it was not measured.
+        """
+        states = [None] * 4
+
+        for a in _not_none(*data):
+            if a.spin_state is SpinChannel.DOWNDOWN:
+                states[0] = a
+            elif a.spin_state is SpinChannel.DOWNUP:
+                states[1] = a
+            elif a.spin_state is SpinChannel.UPDOWN:
+                states[2] = a
+            elif a.spin_state is SpinChannel.UPUP:
+                states[3] = a
+
+        return states
+
+    def process_beams(self, reduction_options=None):
+
+        if reduction_options is None:
+            reduction_options = {
+                "lo_wavelength": 2.5,
+                "hi_wavelength": 12.5,
+                "rebin_percent": 3,
+            }
+
+        for beam in _not_none(self.mm, self.mp, self.pm, self.pp):
```

Comment:
Hmm, what happens if you want to use different options for each state? The wavelength/rebin_percent should be the same, but people might want to specify peak positions, etc, which would require a different reduction options for each spin state.

Reply:
I figure in this case, the user can process each spin state manually using something like `spinset.dd.process(**reduction_options_dd)`. I think it makes more sense for `process_beams` to assume identical options for processing rather than having to worry about the order of a list of `ReductionOptions` passed to it

Reply:
The different beams have to be processed onto the same wavelength axis, which is the benefit of using identical options. However, I was wondering about what happens during location of the specular peak. For the NSF channels the auto-peak location may work really well, but for the SF channels it may prove troublesome.

Thus, one may want to use a mix of automated peak finding and manual beam finding for different spin channels.

Reply:
What about adding `manual_beam_find : ManualBeamFinder()` and `peak_pos : -1` to `reduction_options` if a SF channel is present?

Reply:
I seem to be having trouble using ManualBeamFinder lately, it makes Python stop working...

Reply:
How are you using it? In Jupyter:

```
%gui qt
# IMPORTANT: wait for a few seconds

from refnx.reduce.manual_beam_finder import ManualBeamFinder

mbf = ManualBeamFinder()
```

Reply:
I am sure I had that (including %gui qt) the other day and was still having problems, but seems to be working now. All good.

Reply:
Hi Andy, I have now added reduction options as attributes in SpinSet for each spin channel in one of the latest commits. There are some failed tests but from my understanding they don't seem to be because of my code (maybe an outdated package dependency?).

Reply:
Hi Andy, I have now added reduction options as attributes in the SpinSet class for each spin channel. The user can create the SpinSet object, then update the individual reduction options as they please, then process the spectra with those options. One or two tests have failed but I think this is due to a package dependency or something